examples/sparse/twirls.py

```diff
@@ -0,0 +1,199 @@
+"""
+[Graph Neural Networks Inspired by Classical Iterative Algorithms]
+(https://arxiv.org/pdf/2103.06064.pdf)
+
+This example shows a simplified version of the TWIRLS model proposed
+in the paper. It implements two variants. One is the basic iterative
+graph diffusion algorithm. The other is an advanced implementation
+with attention.
+"""
+
+import argparse
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.optim import Adam
+
+from dgl.data import CoraGraphDataset
+import dgl.mock_sparse as dglsp
+
+
+class MLP(nn.Module):
+    def __init__(self, in_size, hidden_size):
+        super().__init__()
+        self.linear_1 = nn.Linear(in_size, hidden_size)
+        self.linear_2 = nn.Linear(hidden_size, hidden_size)
+        self.dropout = nn.Dropout(0.8)
+
+    def forward(self, X):
+        H = self.linear_1(X)
+        H = F.relu(H)
+        H = self.dropout(H)
+        H = self.linear_2(H)
+        return H
+
+
+################################################################################
+# (HIGHLIGHT) Use DGL sparse API to implement the iterative graph diffusion
+# algorithm.
+################################################################################
+class TWIRLS(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        hidden_size=128,
+        num_steps=16,
+        lam=1.0,
+        alpha=0.5,
+    ):
+        super().__init__()
+        self.num_steps = num_steps
+        self.lam = lam
+        self.alpha = alpha
+        self.mlp = MLP(in_size, hidden_size)
+        self.linear_out = nn.Linear(hidden_size, out_size)
+
+    def forward(self, A, X):
+        # Compute Y = Y0 = f(X; W) using a two-layer MLP.
+        Y = Y0 = self.mlp(X)
+
+        # Compute diagonal matrix D_tild.
+        I = dglsp.identity(A.shape, device=A.device)
+        D_tild = self.lam * dglsp.diag(A.sum(1)) + I
+
+        # Iteratively compute new Y by equation (6) in the paper.
+        for k in range(self.num_steps):
+            Y_hat = self.lam * A @ Y + Y0
+            # The inverse of a diagonal matrix inverses its diagonal values.
+            Y = (1 - self.alpha) * Y + self.alpha * (D_tild**-1) @ Y_hat
```

Comment:
Consider wrapping L61-69 with a function. The same code snippet has appeared for three times.